### PR TITLE
[RW-4773][risk=no] Clean up and organize api/build.gradle a bit

### PR DESCRIPTION
**Pull request by jaycarlton**
The main things I did were:
- alphabetized most of the lists. Tricky b/c of the mix of single and double quotes
- scoped the external symbols so IJ recognized them
- renamed our custom source sets so that it was obvious which Java Plugin tasks were named after them. This is both really nice and somewhat horrifying.
- Added group and description properties to the swagger tasks
- Tried to add plugin notes and links to docs for plugin detritus. 
 
Still to do (possibly on this ticket):
- determine if there's a standard ordering for top level blocks like buildscript, dependencies, version, configurations, etc
- see if it makes sense to organize the file into sections to make a multi-project build easier
- verify that we still actually need all the plugins, tasks, dependencies, and imports

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

**Comment by jaycarlton** [comment]
@gjuggler @ericsong I tried to add you guys as optional reviewers, but either something changed in GitHb or I just couldn't figure out how.

**Comment by jaycarlton** [comment]
I was going to take a bunch of screenshots of what it looks like in IntelliJ, but if you don't mind pulling the branch and poking around yourselves, that's probably more efficient.

A few task names did change, so I need to verify that we don't actually invoke those directly in our processes and make 
Screen Shot 2020-04-15 at 6 03 15 PM

File: api/build.gradle [hunk 1]
```diff
@@ -628,15 +684,15 @@ version = '0.1.0'          // Version in generated output
...
+task incrementalHotSwap(type: IncrementalHotSwapTask, dependsOn: 'compileJava') {
     inputDir = sourceSets.main.java.outputDir
 }
 
-incrementalHotSwap.dependsOn compileJava
-
 task listProjectAPIs(type: GenerateAPIListingTask)
 
 
+// TODO(jaycarlton) move tools-related stuff to its own Gradle project
```
Comment (calbach) [inline api/build.gradle:694]: I don't disagree that this is probably the ideal end state, but I would not attempt to go down this path without first restructuring the codebase holistically to be more modular with cleaner dependencies. Originally, tools were a separate gradle file. This worked to a certain point and then completely broke down as more service dependencies were pulled in. In the end we wound up merging that into this file: https://precisionmedicineinitiative.atlassian.net/browse/RW-2172

File: api/build.gradle [hunk 2]
```diff
@@ -402,36 +464,36 @@ dependencies {
...
   compile 'ch.qos.logback:logback-classic:1.2.3'
 
   compile 'org.springframework.security:spring-security-web:4.2.13.RELEASE'
-  compile "org.hibernate:hibernate-core:${HIBERNATE_VERSION}"
+  compile "org.hibernate:hibernate-core:$project.ext.HIBERNATE_VERSION"
 
   // Force the Kotlin version, otherwise the Kotlin plugin conflicts with Jackson's
   // Kotlin dep, resulting in massive logspam.
-  compile "org.jetbrains.kotlin:kotlin-reflect:${KOTLIN_VERSION}"
+  compile "org.jetbrains.kotlin:kotlin-reflect:$project.ext.KOTLIN_VERSION"
```
Comment (calbach) [inline api/build.gradle:490]: Why the change to omit curly brackets? I'm surprised this is legal

File: api/build.gradle [hunk 3]
```diff
@@ -316,7 +346,38 @@ sourceSets {
...
+
+  //
+  // Custom SourceSets
+  //
+  // Sometimes, in order to fight the clean code war, you have to get dirty in a battle.
+  // Here we use a naming convention with __ word boundaries at start and end
+  // to avoid confusion. For example, the plugin may define a __tools__Compile configuration
+  // or a compile__tools__Java task. This "shields up" approach makes it obvious where there's
+  // a dynamically defined symbol name, so usages stick out and it's easier to identify the
+  // constituent parts. Why four underscores where two might do? Groovy is making me miss C++
```
Comment (calbach) [inline api/build.gradle:358]: nit: Not that I don't like reading this, but I'd save the editorialized first person narration for the PR review and just leave the essentials in the code itself.

File: api/build.gradle [hunk 4]
```diff
@@ -3,13 +3,82 @@ import io.swagger.codegen.DefaultGenerator
...
+    classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$project.ext.KOTLIN_VERSION"
+    classpath "org.jetbrains.kotlin:kotlin-noarg:$project.ext.KOTLIN_VERSION"
+    classpath 'org.owasp:dependency-check-gradle:5.1.0'
+    classpath "org.springframework.boot:spring-boot-gradle-plugin:$project.ext.SPRING_BOOT_VERSION"
+  }
+}
+
+project.ext['hibernate.version'] = project.ext.HIBERNATE_VERSION
+
+// TODO(jaycarlton): migrate to new plugin DSL (RW0-4772)
```
Comment (calbach) [inline api/build.gradle:56]: s/RW0/RW/

File: api/build.gradle [hunk 5]
```diff
@@ -3,13 +3,82 @@ import io.swagger.codegen.DefaultGenerator
...
+// TODO(jaycarlton): move the plugin apply statements closer to configuration blocks for the plugins
```
Comment (calbach) [inline api/build.gradle:57]: Is there a reason not to do this now? Or are you waiting on the new plugin DSL to enable this?

File: api/build.gradle (hunk 1 above)
Comment (jaycarlton) [reply api/build.gradle:694]: Definitely good to know. I'm pulling the TODOs out as they're mostly too prescriptive and the world may have changed before we get to them.

File: api/build.gradle (hunk 2 above)
Comment (jaycarlton) [reply api/build.gradle:490]: IntelliJ suggests that they're redundant and asks to elide them for me.
Screen Shot 2020-04-16 at 9 26 01 AM

From the [docs](https://groovy-lang.org/syntax.html#_string_interpolation)

> In addition to ${} placeholders, we can also use a lone $ sign prefixing a dotted expression:

def person = [name: 'Guillaume', age: 36]
assert "$person.name is $person.age years old" == 'Guillaume is 36 years old'
But only dotted expressions of the form a.b, a.b.c, etc, are valid. Expressions containing parentheses like method calls, curly braces for closures, dots which aren’t part of a property expression or arithmetic operators would be invalid.

File: api/build.gradle (hunk 3 above)
Comment (jaycarlton) [reply api/build.gradle:358]: Yeah, people finding their afternoon taking a detour into the build script generally aren't in the most jocular mood.

By the way, IntelliJ annotates the directories associated with the source sets in the project pane:
Screen Shot 2020-04-16 at 9 32 21 AM

File: api/build.gradle (hunk 5 above)
Comment (jaycarlton) [reply api/build.gradle:57]: The DSL seems to recommend all of them together in a `plugins` block, so doing this wouldn't help with the transition. Killing this one as it's more of a note to self.

File: api/build.gradle [hunk 6]
```diff
@@ -3,13 +3,83 @@ import io.swagger.codegen.DefaultGenerator
...
+    classpath "io.swagger:swagger-codegen:$project.ext.SWAGGER_CODEGEN_VERSION"
+    classpath 'net.ltgt.gradle:gradle-apt-plugin:0.21'
+    classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$project.ext.KOTLIN_VERSION"
+    classpath "org.jetbrains.kotlin:kotlin-noarg:$project.ext.KOTLIN_VERSION"
+    classpath 'org.owasp:dependency-check-gradle:5.1.0'
+    classpath "org.springframework.boot:spring-boot-gradle-plugin:$project.ext.SPRING_BOOT_VERSION"
+  }
+}
+
+project.ext['hibernate.version'] = project.ext.HIBERNATE_VERSION
```
Comment (gjuggler) [inline api/build.gradle:57]: How does this relate to buildscript.ext.HIBERNATE_VERSION? If you have a clear idea of what this line is doing / why it exists, a short comment here might help.

File: api/build.gradle [hunk 7]
```diff
@@ -316,7 +347,36 @@ sourceSets {
     compileClasspath += generated.output
     runtimeClasspath += generated.output
   }
-  integration {
+
+  //
+  // Custom SourceSets
+  //
+  // Here we use a naming convention with __ word boundaries at start and end
```
Comment (gjuggler) [inline api/build.gradle:354]: Nice comment – thanks for taking the time to make this clear.

File: api/build.gradle [hunk 8]
```diff
@@ -3,13 +3,83 @@ import io.swagger.codegen.DefaultGenerator
 import org.pmiops.workbench.tooling.GenerateAPIListingTask
 import org.pmiops.workbench.tooling.IncrementalHotSwapTask
 
+// Runs before all tasks. Sets up properties and dependencies for the build
+// process itself.
+buildscript {
+  // Project-wide constants.
```
[the remark this answers is not in the document]
Comment (jaycarlton) [reply api/build.gradle:9]: I'm confused too. I don't think it makes sense to have an `ext` block inside a `buildscript`.

For example, when I do `./gradlew properties`, these are at the top level with the others:

```
jaycarlton~/repos/workbench/api (jaycarlton/cleanGradle) $ ./gradlew properties

> Task :properties 

------------------------------------------------------------
Root project
------------------------------------------------------------

BootRepackage: class org.springframework.boot.gradle.repackage.RepackageTask
ELASTICSEARCH_VERSION: 6.8.3
GAE_VERSION: 1.9.64
GSON_VERSION: 2.8.5
GenerateReDoc: class org.hidetake.gradle.swagger.generator.GenerateReDoc
GenerateSwaggerCode: class org.hidetake.gradle.swagger.generator.GenerateSwaggerCode
GenerateSwaggerUI: class org.hidetake.gradle.swagger.generator.GenerateSwaggerUI
HIBERNATE_VERSION: 5.2.10.Final
JACKSON_DATABIND_VERSION: 2.9.10.1
JACKSON_VERSION: 2.9.10
JODA_VERSION: 2.10
KOTLIN_VERSION: 1.3.50
MAPSTRUCT_VERSION: 1.3.1.Final
MOCKITO_KOTLIN_VERSION: 2.2.0
OKHTTP_VERSION: 2.7.5
OPENCENSUS_VERSION: 0.23.0
SPRINGFOX_VERSION: 2.6.1
SPRING_BOOT_VERSION: 1.5.22.RELEASE
SWAGGER_CODEGEN_VERSION: 2.2.3
# ...
```

File: api/build.gradle [hunk 9]
```diff
@@ -337,219 +398,223 @@ sourceSets {
...
-}
-
+// TODO(jaycarlton): see if there's a way to subdivide this block for readability
+// To show the dependency tree, try: ./project.rb gradle dependencies --configuration compile
 dependencies {
+  annotationProcessor "org.mapstruct:mapstruct-processor:$project.ext.MAPSTRUCT_VERSION"
 
-  // To show the dependency tree, try: ./project.rb gradle dependencies --configuration compile
+  compile "ch.qos.logback:logback-classic:1.2.3"
+  // Force these dependencies to safe versions, Zendesk client depends on this.
```
Comment (calbach) [inline api/build.gradle:457]: contextual placement of this comment was lost here, since it now seems to refer to all following. Delete if you don't intend to maintain a semantic grouping of dependencies

Comment (jaycarlton) [reply api/build.gradle:457]: I made three copies, but, yeah. I don't think it'll be missed.
